File: cliente-p2p/Infrastructure Layer/Transport/src/main/java/transport/TcpTransportStrategy.java
package transport;

import java.io.*;
import java.net.Socket;
import transport.interfaces.ITransportStrategy;

    public class TcpTransportStrategy implements ITransportStrategy {

    private final String host;
    private final int port;

    public TcpTransportStrategy(String host, int port) {
        this.host = host;
        this.port = port;
    }

    @Override
    public String sendJson(String jsonToSend) {
        System.out.println("[DEBUG] Intentando conectar a " + host + ":" + port);

        try (Socket socket = new Socket(host, port)) {
            System.out.println("[DEBUG] Conexión establecida.");

            // Enviar el JSON
            OutputStream output = socket.getOutputStream();
            PrintWriter writer = new PrintWriter(new OutputStreamWriter(output), true);
            System.out.println("[DEBUG] Enviando JSON: " + jsonToSend);
            writer.println(jsonToSend);

            // Leer la respuesta
            InputStream input = socket.getInputStream();
            BufferedReader reader = new BufferedReader(new InputStreamReader(input));
            String response = reader.readLine();

            System.out.println("[DEBUG] Respuesta recibida: " + response);

            return response;

        } catch (IOException e) {
            System.err.println("[ERROR] Error durante la comunicación: " + e.getMessage());
            e.printStackTrace();
            return null;
        }
    }
}

File: server/InterfaceLayer/Transport/TcpTransportStrategy.go
package transport

import (
    "fmt"
    "io"
    "net"
    "bufio"
    "strings"
)

// TCPTransportStrategy implementa la interfaz ITransportStrategy
type TCPTransportStrategy struct {
    Host string
    Port int
}

// Asegura que TCPTransportStrategy implementa ITransportStrategy
var _ ITransportStrategy = (*TCPTransportStrategy)(nil)

// NewTCPTransportStrategy crea una nueva instancia de TCPTransportStrategy
func NewTCPTransportStrategy(host string, port int) *TCPTransportStrategy {
    return &TCPTransportStrategy{
        Host: host,
        Port: port,
    }
}

// SendJson envía un JSON a través de TCP y recibe la respuesta
func (t *TCPTransportStrategy) SendJson(jsonToSend string) string {
    fmt.Println("[DEBUG] Intentando conectar a", t.Host, ":", t.Port)

    // Conectar al servidor
    conn, err := net.Dial("tcp", fmt.Sprintf("%s:%d", t.Host, t.Port))
    if err != nil {
        fmt.Println("[ERROR] Error durante la conexión:", err)
        return ""
    }
    defer conn.Close()

    fmt.Println("[DEBUG] Conexión establecida.")

    // Enviar el JSON
    fmt.Println("[DEBUG] Enviando JSON:", jsonToSend)
    _, err = fmt.Fprintln(conn, jsonToSend)
    if err != nil {
        fmt.Println("[ERROR] Error enviando el JSON:", err)
        return ""
    }

    // Leer la respuesta
    reader := bufio.NewReader(conn)
    response, err := reader.ReadString('\n')
    if err != nil && err != io.EOF {
        fmt.Println("[ERROR] Error leyendo la respuesta:", err)
        return ""
    }

    // Limpiar el salto de línea
    response = strings.TrimSpace(response)

    fmt.Println("[DEBUG] Respuesta recibida:", response)

    return response
}
